File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    // scan the preorder tree, for each node in the array:
        // find it in inorder array 
        // construct the current node first
        // construct the left sub-tree
        // construct the right sub-tree
    // we need connections between two input array

    var map = {};
    var i = 0;
    inorder.forEach((item, index) => {
        map[item] = index;
    });
    // 9: 0
    // 3: 1
    // 15: 2

    var constructor = function(left, right) {
        var currentNodeValue = preorder[i];
        // preorder build
        var currentNodeIndex = map[currentNodeValue];
        var currentNode = new TreeNode(currentNodeValue);
        if(left < currentNodeIndex) {
            i++;
            currentNode.left = constructor(left, currentNodeIndex - 1);
        }
        if(right > currentNodeIndex) {
            i++;
            currentNode.right = constructor(currentNodeIndex + 1, right);
        }
        return currentNode;
    }

    return constructor(0, inorder.length - 1);
};

File: 0105-construct-binary-tree-from-preorder-and-inorder-traversal/0105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode rec(int preIndex,int inStart,int inEnd,int preorder[],int inorder[]){
        if(preIndex>=preorder.length || inStart>inEnd){
            return null;
        }
        TreeNode newnode=new TreeNode(preorder[preIndex]);
        //find left and right subtrees using inorder array
        int index=0;
        for(int i=inStart;i<=inEnd;i++){
            if(inorder[i]==preorder[preIndex]){
                index=i;
                break;
            }
        }
        //now call for left subtree and right subtree
        newnode.left=rec(preIndex+1,inStart,index-1,preorder,inorder);
        newnode.right=rec(preIndex+index-inStart+1,index+1,inEnd,preorder,inorder);
        
        return newnode;
        
    }
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        return rec(0,0,inorder.length-1,preorder,inorder);
    }
}